docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+4. New feature flag promoted and eventually removed.
+5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.
+
+### Scaling Clusters with No Monitoring Stack
+Monitoring stack is not deployed on clusters with `testing` purpose. An alternative mode of autoscaling will be used for
+such clusters: In the short term, they will be scaled via the existing "class-based scaling + HPA" mode, which is the
+current alternative to HVPA.
+
+### Vision Beyond This Enhancement
+This proposal is part of a mid-term vision which:
+1. Utilizes a different approach for scaling ShootKapi on `testing` clusters, to account for the absence of monitoring
+   stack on them.
+2. For any shoot, enables the user to select an alternative ShootKapi scaling mode which maximises reliability and is
+   known to work well even under extreme API request profiles.
```

Comment:
Our goal should be to introduce a solution that works well across the board.
Mentioning in this GEP that another mode will be introduced later on that is better for certain cases, questions the feasibility of the proposal for me.

Additionally, users should not be able to configure such details as API server requests/limits. This increases the API surface unnecessarily and allows them to intervene too deeply regarding seed resource usage.

Reply:
> Additionally, users should not be able to configure such details as API server requests/limits. This increases the API surface unnecessarily and allows them to intervene too deeply regarding seed resource usage.

I was absolutely arguing in the same direction, until I found out that this is what happens already today, but with worse means. Currently everything needs to be done by operators, because Cluster owners can't do this themselves. And currently we have to take measures to prevent the next reconciliation overwriting these changes.

> Our goal should be to introduce a solution that works well across the board.

In an ideal world, you're right. However, no matter how good your autoscaling approach is, it will always be limited by the size of an individual node and the mechanism preventing that the vertical scale of an apiserver will be so large that it no longer fits on a node: we introduced a `maxAllowed` value for all critical items a while back.
This is by no means the best solution, and we could think about complicating this a bit more by trying to find out reasonable `maxAllowed` values dynamically based on the used worker sizes – but currently we don't do this and I'm not sure if there's a good way to do this.
So for now, the only thing that you can do in these cases is manual intervention: choose bigger controlplane machines, put the apiservers on them, adapt the `maxAllowed` and set manual values for requests/limits/replicas.

Not offering this as a first-class option doesn't prevent it from happening – it only makes these cases more cumbersome to handle for everyone.

Reply:
I think that @voelzmo's comment covers the most important points. I only wanted to add that:
>will be introduced later

No, not later. Not in the sense you're using it. The currently existing fallback already has the property "maximises reliability and is known to work well even under extreme API request profiles". The part of the GEP to which you are referring, says that the new approach will have this property, not that it will be the first approach to have this property.